x/mongo/driver/operation/hello_test.go

```diff
@@ -0,0 +1,1071 @@
+package operation
+
+import (
+	"runtime"
+	"testing"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/internal/assert"
+	"go.mongodb.org/mongo-driver/internal/require"
+	"go.mongodb.org/mongo-driver/version"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
```

Comment:
Suggest testing which parts of the `client` document are omitted when hitting the size limit. Use `t.Setenv` to override environment variables.

[This branch](https://github.com/kevinAlbs/mongo-go-driver/tree/retry_append) has an implementation. Here is the relevant part:

```go
func TestEncodeClientMetadata(t *testing.T) {
	// Clear environment variables that influence handshake.
	clearEnv(t)

	// Set environment variables to add `env` field to handshake.
	t.Setenv(envVarAWSLambdaRuntimeAPI, "lambda")
	t.Setenv(envVarAWSLambdaFunctionMemorySize, "123")
	t.Setenv(envVarAWSRegion, "region")

	// Test that nothing is omitted.
	{
		got, err := encodeClientMetadata("foo", 512)
		assert.Nil(t, err, "error in encodeClientMetadata: %v", err)

		expectD := bson.D{
			{"application", bson.D{
				{"name", "foo"},
			}},
			{"driver", bson.D{
				{"name", driverName},
				{"version", version.Driver},
			}},
			{"os", bson.D{
				{"type", runtime.GOOS},
				{"architecture", runtime.GOARCH},
			}},
			{"platform", runtime.Version()},
			{"env", bson.D{
				{"name", "aws.lambda"},
				{"memory_mb", 123},
				{"region", "region"},
			}},
		}
		expect, err := bson.Marshal(expectD)
		assert.Nil(t, err, "error in Marshal: %v", err)
		assert.Equal(t, expect, got, "expected: %s, but got %s", bson.Raw(expect), bson.Raw(got))
	}

	// Test that fields from `env`` are omitted first.
	{
		// Reduce max size to 225.
		got, err := encodeClientMetadata("foo", 225)
		assert.Nil(t, err, "error in encodeClientMetadata: %v", err)

		expectD := bson.D{
			{"application", bson.D{
				{"name", "foo"},
			}},
			{"driver", bson.D{
				{"name", driverName},
				{"version", version.Driver},
			}},
			{"os", bson.D{
				{"type", runtime.GOOS},
				{"architecture", runtime.GOARCH},
			}},
			{"platform", runtime.Version()},
			{"env", bson.D{
				{"name", "aws.lambda"},
			}},
		}
		expect, err := bson.Marshal(expectD)
		assert.Nil(t, err, "error in Marshal: %v", err)
		assert.Equal(t, expect, got, "expected: %s, but got %s", bson.Raw(expect), bson.Raw(got))
	}

	// Test that entire env document is omitted next.
	{
		// TODO.
	}

	// Test that fields from os are omitted next.
	{
		// TODO.
	}

	// Test that platform is truncated next.
	{
		// TODO.
	}

	// Test that empty slice is returned last.
	{
		// TODO.
	}
}
```